### ISSUE TITLE
F* blows up: bound term variable not found

### ISSUE THREAD

**Issue by catalin-hritcu** [FStarLang/FStar#170]
```
assume val sorted: list int -> Tot bool
assume val split: l:list int -> Tot ((list int * list int))
assume val merge: l1:list int -> Tot (r:list int{(sorted l1)})

val mergesort: l:list int -> list int
let rec mergesort l =
    let (l1, l2) = split l in
    let sl1 = mergesort l in
    merge sl1
```

**Comment by catalin-hritcu** [comment]
Here is a stack-trace for this (BTW: where is the flag for getting a stack trace, needed to use the OCaml version to get this):
Unexpected error Failure("Bound term variable not found: sl1"):
Raised at file "pervasives.ml", line 30, characters 22-33
Called from file "Microsoft_FStar_ToSMT_Encode.ml", line 790, characters 1-24
Called from file "Microsoft_FStar_ToSMT_Encode.ml", line 1116, characters 14-32
Called from file "list.ml", line 84, characters 24-34
Called from file "Microsoft_FStar_ToSMT_Encode.ml", line 883, characters 14-38
Called from file "Microsoft_FStar_ToSMT_Encode.ml", line 1230, characters 14-32
Called from file "support.ml", line 387, characters 25-34
Called from file "list.ml", line 84, characters 24-34
Called from file "support.ml", line 388, characters 26-62
Called from file "Microsoft_FStar_ToSMT_Encode.ml", line 1219, characters 86-689
Called from file "Microsoft_FStar_ToSMT_Encode.ml", line 1127, characters 51-87
Called from file "Microsoft_FStar_ToSMT_Encode.ml", line 665, characters 17-40
Called from file "Microsoft_FStar_ToSMT_Encode.ml", line 1109, characters 14-37
Called from file "list.ml", line 84, characters 24-34
Called from file "Microsoft_FStar_ToSMT_Encode.ml", line 726, characters 14-36
Called from file "Microsoft_FStar_ToSMT_Encode.ml", line 578, characters 14-37
Called from file "Microsoft_FStar_ToSMT_Encode.ml", line 547, characters 14-63
Called from file "list.ml", line 84, characters 24-34
Called from file "Microsoft_FStar_ToSMT_Encode.ml", line 1359, characters 61-98
Called from file "Microsoft_FStar_ToSMT_Encode.ml", line 1418, characters 17-51
Called from file "src/batList.ml", line 517, characters 26-35
Called from file "src/batList.ml", line 521, characters 2-18
Called from file "Microsoft_FStar_ToSMT_Encode.ml", line 1241, characters 81-368
Called from file "Microsoft_FStar_ToSMT_Encode.ml", line 1377, characters 14-51
Called from file "Microsoft_FStar_ToSMT_Encode.ml", line 1418, characters 17-51
Called from file "src/batList.ml", line 517, characters 26-35
Called from file "src/batList.ml", line 521, characters 2-18
Called from file "Microsoft_FStar_ToSMT_Encode.ml", line 1241, characters 81-368
Called from file "Microsoft_FStar_ToSMT_Encode.ml", line 2490, characters 17-60
Called from file "Microsoft_FStar_Tc_Util.ml", line 6, characters 36-86
Called from file "Microsoft_FStar_Tc_Tc.ml", line 1959, characters 47-61
Called from file "Microsoft_FStar_Tc_Tc.ml", line 1690, characters 17-38
Called from file "src/batList.ml", line 171, characters 23-28
Called from file "Microsoft_FStar_Tc_Tc.ml", line 2247, characters 24-1023
Called from file "Microsoft_FStar_Tc_Tc.ml", line 2384, characters 17-46
Called from file "list.ml", line 84, characters 24-34
Called from file "Microsoft_FStar_Tc_Tc.ml", line 2378, characters 69-1023
Called from file "Microsoft_FStar_Tc_Tc.ml", line 2475, characters 17-135
Called from file "Microsoft_FStar_Tc_Tc.ml", line 2514, characters 14-30
Called from file "list.ml", line 84, characters 24-34
Called from file "fstarml.ml", line 80, characters 21-69
Called from file "fstarml.ml", line 145, characters 32-37
Called from file "support.ml", line 15, characters 27-32

**Comment by catalin-hritcu** [comment]
There were a couple more instances of the bug that are fixed now (one in #166 and another one in https://github.com/FStarLang/FStar/blob/master/examples/metatheory/norm-take2.fst).

**Comment by catalin-hritcu** [comment]
Marking this as fixed, there is still #207 though.
